symengine/parser/tokenizer.re

```diff
@@ -0,0 +1,176 @@
+    // Solution: Increase SIZE.
+    // Note: Whenever the buffer fills, the data (all the loaded tokens in the
+    // buffer) gets moved in it and we start over. To ensure this happens
+    // rarely, make SIZE large enough, for example 64*1024 (64 KB). The amount
+    // of data to move is typically low, so this size should be large enough
+    // not to be noticeable. One can run a benchmark study what size stops
+    // being noticeable for a large file.
+    ERR_BUF,
+
+    // We reached the end of input.
+    END,
+    WS, OPERATOR, POW, LE, EQ, GE, IDENTIFIER, NUMERIC,
+    IMPLICIT_MUL};
+
+/*!max:re2c*/
+static const size_t SIZE = 64*1024;
+
+size_t sread(char *ptr, size_t size, size_t count, std::istream &stream)
+{
+    stream.read(ptr, count);
+    return stream.gcount();
+}
+
+using SymEngine::Parser;
+
+
+struct input_t {
+    unsigned char buf[SIZE + YYMAXFILL];
+    unsigned char *lim;
+    unsigned char *cur;
+    unsigned char *mar;
+    unsigned char *tok;
+    bool eof;
+
+    std::istream &file;
+
+    SymEngine::ParserBase::STYPE__ *val;
+
+    input_t(std::istream &f)
+        : buf()
+        , lim(buf + SIZE)
+        , cur(lim)
+        , mar(lim)
+        , tok(lim)
+        , eof(false)
+        , file(f)
+    {}
+    bool fill(size_t need)
+    {
+        if (eof) {
+            return false;
+        }
+        const size_t free = tok - buf;
+        if (free < need) {
+            return false;
+        }
+        memmove(buf, tok, lim - tok);
+        lim -= free;
+        cur -= free;
+        mar -= free;
+        tok -= free;
+        lim += sread((char*)lim, 1, free, file);
+        if (lim < buf + SIZE) {
+            eof = true;
+            memset(lim, 0, YYMAXFILL);
+            lim += YYMAXFILL;
+        }
+        return true;
+    }
+};
+
+int yylex(input_t &in)
+{
+    for (;;) {
+        in.tok = in.cur;
+        /*!re2c
+            re2c:define:YYCURSOR = in.cur;
+            re2c:define:YYMARKER = in.mar;
+            re2c:define:YYLIMIT = in.lim;
+            re2c:yyfill:enable = 1;
+            re2c:define:YYFILL = "if (!in.fill(@@)) return 0;";
+            re2c:define:YYFILL:naked = 1;
+            re2c:define:YYCTYPE = "unsigned char";
+
+            end = "\x00";
+
+
+            whitespace = [ \t\v\n\r]+;
+            dig = [0-9];
+            char =  [\x80-\xff] | [a-zA-Z_];
+            operators = "-"|"+"|"/"|"("|")"|"*"|","|"^"|"~"|"<"|">"|"&"|"|";
+
+            pows = "**"|"@";
+            le = "<=";
+            ge = ">=";
+            eqs = "==";
+            ident = char (char | dig)*;
+            numeric = (dig*"."?dig+([eE][-+]?dig+)?) | (dig+".");
+            implicitmul = numeric ident;
+
+            *   { printf("ERR unknown token\n"); return 0; }
+            end {
+                    if (in.lim - in.tok == YYMAXFILL) {
+                        return 0;
+                    } else {
+                        printf("ERR NULL.\n");
```

Comment:
These two print statements are the only ugly parts of the code. Essentially they signal an error. I think we should raise an exception instead?

Reply:
Yes, a `SymEngine::ParseError`.

Reply:
My understanding that both of these error conditions will disappear after I rework the tokenizer per my comment https://github.com/symengine/symengine/pull/1586#issuecomment-517745895.